### PR TITLE
feat(rloo): add synchronous training baseline

### LINKED ISSUE #176: [Task 28] Add Synchronous RLOO Training
## Background

Relax currently supports GRPO and several related policy-gradient estimators, but it does not expose REINFORCE Leave-One-Out (RLOO) as a public baseline. RLOO samples multiple responses for the same prompt and uses the other responses in that prompt group as an unbiased, parameter-free baseline. This reduces policy-gradient variance without adding a critic model.

The implementation should follow *Back to Basics: Revisiting REINFORCE Style Optimization for Learning from Human Feedback in LLMs* (ACL 2024, arXiv:2402.14740) and integrate with Relax's existing synchronous colocate training path.

## Objective

Add `rloo` to the algorithm registry and implement its reward baseline, sequence advantage, policy loss, and monitoring metrics for synchronous training only.

For a prompt group with `K` sampled responses and finite scalar task rewards `r_1, ..., r_K`, define:

```text
KL_i = sum_t m_i,t * (log pi_old(y_i,t | ...) - log pi_ref(y_i,t | ...))
R_i = r_i - beta * KL_i
b_i = (sum_j R_j - R_i) / (K - 1)
A_i = R_i - b_i
L_i = -stop_gradient(A_i) * sum_t m_i,t * log pi_theta(y_i,t | x_i, y_i,<t)
```

where `m_i,t` is the response loss mask. The group objective is the mean of the per-sequence losses. The KL-shaped reward is configured through `--kl-coef`; `--use-kl-loss` is rejected for RLOO to avoid applying the KL penalty twice.

## Scope

- Register `rloo` with the same synchronous actor/rollout role topology used by critic-free policy-gradient algorithms.
- Accept `--advantage-estimator rloo` and reject `--fully-async` and hybrid-async modes for this estimator.
- Require `--n-samples-per-prompt >= 2`.
- Compute sequence KL after the old-policy/reference forwards, gather shaped rewards and `Sample.group_index` across the explicit DP group, then compute each leave-one-out baseline globally. Do not rely on samples being contiguous or colocated on one DP rank.
- Broadcast each scalar sequence advantage to the valid response-token shape used by the Megatron loss path.
- Implement the unclipped sequence-level REINFORCE objective. Existing PPO/GRPO/GSPO/SAPO/CISPO loss behavior must remain unchanged.
- Preserve existing KL metrics and expose RLOO's sequence KL/reward-shaping metrics.
- Report at least:
  - `rloo_shaped_reward`
  - `rloo_sequence_kl`
  - `rloo_baseline`
  - `rloo_advantage`
  - `rloo_advantage_abs`
  - `rloo_empty_response_fraction`
  - existing raw reward, loss, policy-gradient loss, and policy KL metrics
- Add a synchronous example launch script and a reproducible GRPO/RLOO comparison report.

### Group size

- `K < 2` is invalid and must raise a readable configuration error before services are launched.
- Every observed `group_index` must contain exactly `--n-samples-per-prompt` samples. Missing or oversized groups must raise `ValueError` and identify the group and observed size.

### Padding and masks

- Prompt tokens and padding tokens never contribute to the RLOO policy loss.
- Masked response tokens do not contribute to the loss or its denominator.
- Static CP shards first reconstruct the full masked sequence KL/log-probability. Only CP rank 0 contributes the replicated sequence loss, with metric scaling that preserves the unsplit value.

### Empty responses

- A response with zero valid response tokens is accepted as rollout data but contributes zero policy loss and zero gradient.
- Its scalar reward still participates in the leave-one-out baselines of the other samples in the prompt group.
- Empty-response frequency is exposed as a monitoring metric.

### Invalid rewards

- `NaN`, `+inf`, and `-inf` task rewards are rejected before baseline computation.
- The error must identify the sample position or prompt group. Invalid values must not be silently clamped or included in metrics.

## Implementation Notes

Likely integration points:

- `relax/core/registry.py`: register the `rloo` role graph.
- `relax/utils/arguments.py`: parser choice and synchronous/group-size validation.
- `relax/utils/utils.py`: carry stable prompt-group IDs and validate rollout rewards/groups.
- `relax/utils/training/ppo_utils.py`: pure baseline/advantage and element-wise loss helpers.
- `relax/backends/megatron/loss.py`: CP sequence aggregation, explicit DP gather, advantage broadcast, loss, and metrics.
- `relax/utils/training/data_fields.py`: carry RLOO monitoring fields through TransferQueue.
- `tests/`: pure formula tests plus integration-focused mask, DP, and CP tests.
- `examples/algorithms/`: synchronous RLOO launch script.

The independent `Advantages` Ray Serve deployment is not part of this task because fully asynchronous RLOO is explicitly out of scope.

### Formula tests

Use hand-constructed prompt groups with non-contiguous `group_index` values and compare every element against an independently written reference:

```text
group A rewards: [1.0, 0.0, 0.5]
baseline:        [0.25, 0.75, 0.50]
advantage:       [0.75, -0.75, 0.00]
```

Verify the policy loss from explicit token log-probabilities and response masks, including mixed response lengths and padded tokens.

### Boundary tests

- `K = 1` configuration rejection.
- Incomplete and oversized prompt groups.
- Zero-length response and all-zero response mask.
- `NaN`, `+inf`, and `-inf` rewards.
- Float32 and float64 formula consistency.

### Distributed invariance tests

- Split prompt groups across two DP ranks, gather shaped rewards/group IDs with the explicit DP process group, and verify the resulting local baselines/advantages match the unpartitioned reference.
- Split response tokens into simulated CP shards with padding/masks, reduce the shard losses, and verify equality with the unsplit loss on valid tokens.
- Verify sequence-count normalization. Static CP may require the Megatron Bridge `--calculate-per-token-loss` compatibility flag, but RLOO must still reduce gradients and metrics by sequence count rather than valid-token count.

### Regression tests

- Existing default `grpo` parser behavior and role graph are unchanged.
- Existing GRPO return, policy loss, KL, and reward-normalization tests continue to pass.
- `rloo` appears in the registry and resolves to the synchronous colocate role set.
- `rloo --fully-async` fails fast with a readable message.

## Two-GPU Experiment

Run matched experiments on one host with 2 x RTX 5090 (32 GB each), using Qwen3-0.6B and GSM8K to keep the validation affordable.

Hold constant across both runs:

- initial checkpoint and reference checkpoint
- dataset and prompt order/seed
- `rollout_batch_size`, `n_samples_per_prompt`, and `global_batch_size`
- rollout temperature and maximum response length
- optimizer, learning rate, KL coefficient, and number of rollout steps
- actor/rollout placement and SGLang memory fraction

Change only `--advantage-estimator {grpo,rloo}` and algorithm-specific reward normalization. Use at least 10 completed train steps for a smoke comparison; prefer 50-100 steps for a directional training comparison.

Report the same step window for both runs:

| Category | Metrics |
| --- | --- |
| Reward | mean `rollout/raw_reward`, final-window mean, optional eval pass@1 |
| Loss | mean and standard deviation of `train/pg_loss` and `train/loss` |
| KL | mean policy KL for both runs, plus RLOO sequence KL and GRPO explicit KL loss |
| Throughput | samples/s, response tokens/s, actor train step time, rollout time, end-to-end wall time |
| Stability | grad norm, non-finite metric count, OOM/restart count |

The report must state that a short two-GPU run validates execution and metric wiring but is not sufficient to establish algorithmic superiority.

## Acceptance Criteria

- Hand-constructed batches produce element-wise baselines, advantages, and losses matching the formulas above.
- Group size, incomplete groups, padding, empty responses, and non-finite rewards have the explicit behavior documented above.
- DP and CP partitioning do not change valid-token statistics or the reduced loss.
- A small-model synchronous RLOO training run completes on 2 x RTX 5090 and is compared with a matched GRPO run for reward, loss, KL, and throughput.
- The default algorithm remains `grpo`, and existing algorithms' behavior is unchanged.
- Fully asynchronous RLOO is rejected and remains out of scope.
- Focused tests pass and `pre-commit run --all-files` is clean, excluding hardware-dependent multi-node integration tests with an explicit reason.

## Non-Goals

- Fully asynchronous or hybrid-async RLOO.
- A critic/value model.
- New dependencies.
- Changes to Controller, Service, or launcher behavior beyond selecting the registered synchronous role graph.
- Claims of statistically significant RLOO-versus-GRPO quality from a short smoke run.

## References

- Ahmadian et al., *Back to Basics: Revisiting REINFORCE Style Optimization for Learning from Human Feedback in LLMs*, ACL 2024: https://arxiv.org/abs/2402.14740
- Hugging Face TRL RLOO documentation: https://huggingface.co/docs/trl/rloo_trainer

### PR DESCRIPTION
**Pull request by Dayuxiaoshui**
## What

Add synchronous REINFORCE Leave-One-Out (RLOO) as a public, critic-free algorithm.

This change includes:

- An `rloo` registry entry and argument validation.
- Leave-one-out baseline, advantage, sampled KL reward shaping, and sequence-level policy loss.
- Explicit prompt-group IDs and data-parallel aggregation.
- Static context-parallel reconstruction for valid response-token statistics.
- RLOO monitoring metrics and a reproducible two-GPU RLOO/GRPO example.
- Formula, edge-case, DP, CP, model-mask, and metric-aggregation tests.

Asynchronous RLOO is intentionally out of scope.

## Why

Relax provides GRPO and related critic-free algorithms, but it does not provide RLOO as a public baseline. GRPO's normalized group advantages and clipped token objective are not equivalent to RLOO's leave-one-out baseline and sequence-level REINFORCE objective.

A native implementation makes matched RLOO/GRPO comparisons possible without custom reward hooks or a value model.

Fixes #176.

## How

For each prompt group of size $K$ (the $K$ responses generated from the same prompt), RLOO computes a per-sequence KL penalty, a leave-one-out baseline, and a REINFORCE-style policy loss.

### Objective

For each response $i$ in the group:

$$
\text{KL}_i = \sum_t m_{i,t} \left( \log \pi_{\text{old},i,t} - \log \pi_{\text{ref},i,t} \right)
$$

$$
R_i = r_i - \beta \cdot \text{KL}_i
$$

$$
\text{baseline}_i = \frac{\sum_{j} R_j - R_i}{K - 1}
$$

$$
\text{advantage}_i = R_i - \text{baseline}_i
$$

$$
\mathcal{L}_i = -\mathrm{sg}\left( \text{advantage}_i \right) \cdot \sum_t m_{i,t} \log \pi_{\theta,i,t}
$$

$$
\mathcal{L} = \frac{1}{K} \sum_i \mathcal{L}_i
$$

### Notation

| Symbol | Meaning |
| --- | --- |
| $i$ | Index of a response in the prompt group, $i = 1, \dots, K$ |
| $t$ | Token position in response $i$ |
| $m_{i,t}$ | Binary mask: $1$ for real response tokens, $0$ for prompt and padding tokens |
| $r_i$ | Scalar reward for response $i$ from the task or reward model |
| $\beta$ | KL coefficient that controls deviation from the reference policy |
| $\pi_{\text{old}}$ | Policy used to collect the rollout data (old / behavior policy) |
| $\pi_{\text{ref}}$ | Reference policy (typically the SFT checkpoint) |
| $\pi_\theta$ | Current policy being trained |
| $\mathrm{sg}(\cdot)$ | Stop-gradient: the value is treated as a constant during back-propagation |

### Step-by-step explanation

1. **Per-sequence KL estimate** — $\text{KL}_i$ is the masked sum of log-probability differences between the old policy and the reference policy. It measures how far the current rollout distribution has drifted from the reference.

2. **Shaped reward** — $R_i$ subtracts a KL penalty $\beta \cdot \text{KL}_i$ from the raw reward $r_i$, discouraging the policy from moving too far away from the reference.

3. **Leave-one-out baseline** — $\text{baseline}_i$ is the average shaped reward of all responses in the same group except response $i$. This is the key RLOO baseline: each response is compared against the mean of its peers, rather than the group mean.

4. **Advantage** — $\text{advantage}_i$ is how much better (or worse) response $i$ is relative to the leave-one-out baseline. This naturally has mean zero over the group.

5. **Policy loss** — $\mathcal{L}_i$ scales the negative log-likelihood of the response by the stopped-gradient advantage. Stopping the gradient on the advantage keeps the baseline and reward terms from creating unwanted gradients through the policy.

6. **Group average** — $\mathcal{L}$ averages the per-sequence losses over the full group of $K$ responses, giving the final training objective.

### Implementation details

The implementation:

- Requires complete prompt groups with `K >= 2` and rejects non-finite rewards.
- Excludes prompt and padding tokens from KL and policy loss via the mask $m_{i,t}$.
- Gives empty responses zero policy loss and gradient while retaining their scalar reward in peer baselines.
- Gathers group IDs and shaped rewards over the explicit DP group before leave-one-out reduction.
- Reconstructs full response-token log probabilities under static CP before sequence reductions.
- Uses sequence-count loss normalization even when Megatron Bridge requires `--calculate-per-token-loss` for static CP compatibility.
- Supplies a native local BSHD causal mask for the RTX 5090 CP correctness path.
- Keeps existing estimator defaults and non-RLOO behavior unchanged.
- Rejects asynchronous, normalized-advantage, double-KL, dynamic-CP, and unsupported custom RLOO combinations before launch.

### Metrics added

The following rollout metrics are added:

- `rloo_shaped_reward` — per-sequence $R_i$ after KL penalty.
- `rloo_sequence_kl` — per-sequence $\text{KL}_i$ estimate.
- `rloo_baseline` — per-sequence leave-one-out baseline.
- `rloo_advantage` — per-sequence advantage.
- `rloo_advantage_abs` — absolute value of the advantage for monitoring.
- `rloo_empty_response_fraction` — fraction of empty responses in the group.

## Testing

- [ ] `pre-commit run --all-files` passes
- [ ] Tests pass (`pytest tests/`)
- [x] New tests added (if applicable)
- [ ] Documentation updated (if applicable)

Focused test command:

```bash
pytest -q \
  tests/utils/training/test_ppo_utils_rloo.py \
  tests/backends/megatron/test_loss_rloo.py \
  tests/backends/megatron/test_model_rloo_cp.py \
  tests/backends/megatron/test_data_rloo_metrics.py \
  tests/core/test_registry_rloo.py
```

Result: `26 passed, 48 warnings in 18.99s`.

Coverage includes hand-computed elementwise baselines, advantages, masked losses and gradients; invalid `K`; incomplete and non-contiguous groups; padding; empty responses; non-finite rewards; simulated DP/CP partitioning; BSHD CP gather metadata; native CP causal masks; and CP-invariant RLOO monitoring.

The full test suite and multi-node integration suite were not run. The available validation host is one node with two RTX 5090 GPUs. Full pre-commit was attempted, but gitleaks could not download Go dependencies from `proxy.golang.org`, and docformatter 1.3.1 is incompatible with this Python 3.12 environment because `lib2to3` is unavailable. All remaining hooks pass with only those two hooks skipped.

### Matched two-GPU experiment

Configuration: Qwen3-0.6B, GSM8K, one node with 2x RTX 5090, DP=2, CP=1, four prompts per rollout, `K=4`, maximum response length 512, seed 1234, KL coefficient 0.01, and 10 training steps.

| Metric | RLOO | GRPO |
| --- | ---: | ---: |
| Raw reward, 10-step mean | 0.031250 | 0.050000 |
| Train loss, 10-step mean | -0.006706 | -8.571e-7 |
| Policy loss, 10-step mean | -0.006706 | -1.863e-9 |
| `ppo_kl`, 10-step mean | 0.000000 | 0.000000 |
| Algorithm KL monitor, 10-step mean | `rloo_sequence_kl` 0.091381 | `kl_loss` -8.557e-5 |
| Actor throughput, 10-step mean (token/s) | 4226.29 | 3653.66 |
| Actor throughput, steady steps 1-9 (token/s) | 4523.54 | 3866.38 |
| Response throughput, 10-step mean (token/s) | 581.43 | 586.18 |
| Response throughput, steady steps 1-9 (token/s) | 638.95 | 643.96 |
| Gradient norm, 10-step mean | 109.17 | 0.6976 |
| Non-finite / OOM / restart count | 0 / 0 / 0 | 0 / 0 / 0 |

The RLOO sequence KL and GRPO KL-loss fields have different definitions and should not be compared numerically as the same estimator. Likewise, RLOO uses a sequence-sum REINFORCE loss while GRPO uses a token-normalized clipped objective, so their loss and gradient-norm scales are not directly comparable.

Steady response throughput differed by -0.78% for RLOO relative to GRPO. This 10-step smoke test validates execution and metric wiring only; it does not establish convergence or algorithmic superiority.

### Static CP smoke test

A full-model `CP=2`, `DP=1` RLOO training step completed successfully on the same 2x RTX 5090 host:

| Metric | Value |
| --- | ---: |
| Raw reward | 0.125000 |
| Train / policy loss | -3.894598 |
| `ppo_kl` | 0.000000 |
| Actor train time | 14.11 s |
| Actor throughput | 644.33 token/s |
| Response throughput | 59.57 token/s |
| Actor train tokens | 9094 |
| OOM / non-finite / restart count | 0 / 0 / 0 |

The run completed actor training, optimizer update, weight synchronization, and clean shutdown. The CP sequence-monitor values in that run exposed a logging-only extra `cp_size` multiplier; the implementation now excludes replicated `rloo_*` sequence metrics from that multiplier, with a focused regression test.

On this Blackwell host, Transformer Engine flash/fused attention backends were unavailable or failed to load. The successful CP path uses Megatron local BSHD attention with the native zig-zag causal mask added here. This is a correctness fallback with quadratic attention memory, not a long-context production recommendation. Hopper hardware is not required for the demonstrated small-model CP smoke test.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] CI/CD or build changes

## Screenshots / Logs

Successful jobs:

- RLOO DP=2, 10 steps: `raysubmit_HiHaDm2UUqDiKjbd`
- GRPO DP=2, 10 steps: `raysubmit_GviTZSAsJuqnasuG`
- RLOO CP=2 full-model smoke: `raysubmit_rg6Y7VGXzxGNZhy4`

All three jobs completed without OOM, non-finite values, or actor restarts

### CODE REVIEW

File: relax/utils/arguments.py [hunk 1]
```diff
@@ -2743,6 +2744,33 @@ def slime_validate_args(args):
     assert not (args.kl_coef != 0 and args.kl_loss_coef != 0), "Only one of kl_coef and kl_loss_coef can be set"
 
     if not is_sft:
+        if args.advantage_estimator == "rloo":
```
Comment (yuanlehome) [inline relax/utils/arguments.py:2747]: [P1] 这里仍接受一轮 rollout 对应多个 optimizer step。外层会对每个 sub-step 调用 `train_one_step` 并执行 `optimizer.step()`；从第二步开始，样本来自更新前的 policy，但 RLOO 直接优化 `-A * sum(log pi_theta)`，没有 behavior-policy ratio，因此会变成未修正的 off-policy 更新。

`--partial-rollout` 也没有被限制；未同时启用 `--mask-offpolicy-in-partial-rollout` 时，recycled prefix 仍可能参与梯度。当前 baseline 建议强制每轮只有一次 optimizer update，并禁用 partial rollout（或至少强制 mask 旧 prefix）；如果要开放这些配置，需要先实现 sequence-level IS。

File: relax/backends/megatron/model.py [hunk 2]
```diff
@@ -183,6 +184,44 @@ def _main_loss_has_tokens(batch: dict) -> bool:
...
+            f"RLOO native context parallelism requires an even local sequence length, got {local_sequence_length}."
+        )
+
+    chunk_size = local_sequence_length // 2
+    global_sequence_length = local_sequence_length * cp_size
+    cp_rank = mpu.get_context_parallel_rank()
+    first = slice(cp_rank * chunk_size, (cp_rank + 1) * chunk_size)
+    second_chunk = 2 * cp_size - cp_rank - 1
+    second = slice(second_chunk * chunk_size, (second_chunk + 1) * chunk_size)
+    full_mask = torch.triu(
```
Comment (yuanlehome) [inline relax/backends/megatron/model.py]: [P1] 这里在每个 CP rank、每次 forward 都先物化完整的 `global_seq_len × global_seq_len` bool mask，然后只取本 rank 的 `global_seq_len / CP` 行。32K 序列时仅这个 bool tensor 就约 1 GiB/rank，`triu` 还可能产生额外临时量，会在 CP 最需要支持的长上下文场景中直接 OOM。

请直接用本 rank 的 zig-zag local query positions 与 global key positions 做比较，构造 `local_seq_len × global_seq_len` mask，避免创建完整方阵；测试也应断言实际 allocation/output shape，而不只是从一个完整参考方阵切片。

File: tests/backends/megatron/test_data_rloo_metrics.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# Copyright (c) 2026 Relax Authors. All Rights Reserved.
+
+from argparse import Namespace
+
+import pytest
+import torch
+
+from relax.backends.megatron import data
```
Comment (yuanlehome) [inline tests/backends/megatron/test_data_rloo_metrics.py]: [P1] 当前 CI 的 Python 3.10/3.11/3.12 job 都在 collection 阶段因这里的无条件 import 报 `ModuleNotFoundError: megatron`。同样的问题还存在于 `test_loss_rloo.py` 和 `test_model_rloo_cp.py`。

请像 `test_registry_rloo.py` 一样，在任何 Megatron 后端 import 之前使用 `pytest.importorskip("megatron.core")`，然后重跑三个 Python 版本的完整 CI。

**Review by yuanlehome — CHANGES_REQUESTED** [review]
建议 REQUEST_CHANGES。这个实现的 sequence-level loss、completion KL shaping 和 DP-global LOO 方向是合理的，但当前仍有三个合并阻塞项：

1. 参数校验没有保证 RLOO 的 on-policy 条件；同一轮 rollout 多次 optimizer update 后，后续 step 会无修正地训练旧 policy 数据，partial rollout 的旧 prefix 也可能参与梯度；
2. native CP fallback 在每个 rank 上先创建完整 S×S causal mask，长上下文下会产生可避免的 OOM；
3. 当前 CI 失败，Python 3.10/3.11/3.12 都在 collection 阶段因新增测试无条件 import Megatron 而退出。

另有两个应在本 PR 处理的次要问题：debug subsampling 没有把 RLOO 纳入 group-aware 分支，会切断 K-response group；`group_indices` 没有从 rollout metrics 中排除，会被错误地求平均上报。

修复后请重新运行完整 CI，并补多-step/partial-rollout 参数测试、完整 group 的 debug subsampling 回归，以及不物化 global S×S mask 的 shape/内存测试。真实多进程 DP/CP 集成若受硬件限制无法运行，请明确记录跳过原因。

**Comment by SigureMo** [comment]
#205 merged，本 PR closed
